### PR TITLE
Support Java 21

### LINKED ISSUE #335: Support Java 21
Currently the project doesn't build using Java 21, let's fix this!

### PR DESCRIPTION
**Pull request by jdubois**
This PR is to fix #335 

It seems that all there needs to do is upgrade Mockito to version 5.8.0.

### CODE REVIEW

**Comment by jdubois** [comment]
Argh I missed you were testing for Java 8 support! It's legacy now, like Java 11! And I'm guessing people using LangChain4J are working on newer projects, that wouldn't require this anyway.

**Comment by langchain4j** [comment]
Surprisingly a lot of people still use java 11, even java 8. We did an effort to stay compatible with java 8 so that more people can take advantage of this library.

**Comment by jdubois** [comment]
Sorry @langchain4j I realised I was also testing on the main project's CI, so I converted this to draft.

**Comment by jdubois** [comment]
With this PR, we migrate to Java 21 (easy, that's only Mockito to upgrade), and change the build:

- For Java 11 and 17, everything is tested excepted GraalVM
- For Java 21, everything is tested

Concerning GraalVM, support for Java 17 will drop in 9 months, and there's no reason to stay on an older version as this creates a native image.

For Java 8, and even 11: those are not supported at all, so you will get CVEs. Those versions also don't work well in Docker. They should not be used anymore.

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -1,39 +1,37 @@
 name: Java CI
 
-on: [pull_request]
+on:
+  push:
+    branches:
+      - main
+      - support-java21
```
Comment (langchain4j) [inline .github/workflows/main.yaml]: `support-java21` can be removed, right?

**Comment by langchain4j** [comment]
@jdubois could you please explain why is it required to upgrade mockito in order to build with java 21? I've just built locally and it seems to work well with 4.11.

**Comment by jdubois** [comment]
@langchain4j when running `mvn test` with Java 21, I'm getting errors from Mockito 4.11.0 :

```java
[ERROR] dev.langchain4j.data.document.S3FileLoaderTest.should_load_document -- Time elapsed: 0.055 s <<< ERROR!
org.mockito.exceptions.base.MockitoException:

Mockito cannot mock this class: interface software.amazon.awssdk.services.s3.S3Client.

If you're not sure why you're getting this error, please open an issue on GitHub.


Java               : 21
JVM vendor name    : Homebrew
JVM vendor version : 21.0.1
JVM name           : OpenJDK 64-Bit Server VM
JVM version        : 21.0.1
JVM info           : mixed mode, sharing
OS name            : Mac OS X
OS version         : 14.1.2


You are seeing this disclaimer because Mockito is configured to create inlined mocks.
You can learn about inline mocks and their limitations under item #39 of the Mockito class javadoc.

Underlying exception : org.mockito.exceptions.base.MockitoException: Could not modify all classes [interface software.amazon.awssdk.awscore.AwsClient, interface software.amazon.awssdk.utils.SdkAutoCloseable, interface java.lang.AutoCloseable, interface software.amazon.awssdk.core.SdkClient, interface software.amazon.awssdk.services.s3.S3Client]
	at org.mockito.junit.jupiter.MockitoExtension.beforeEach(MockitoExtension.java:153)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	Suppressed: java.lang.NullPointerException: Cannot invoke "java.util.Set.forEach(java.util.function.Consumer)" because the return value of "org.junit.jupiter.api.extension.ExtensionContext$Store.remove(Object, java.lang.Class)" is null
		at org.mockito.junit.jupiter.MockitoExtension.afterEach(MockitoExtension.java:184)
		... 2 more

**Comment by jdubois** [comment]
@langchain4j I realise you have a lot of other old dependencies (for example Spring Boot 2.7, which also is unmaintained now!). Have you had a look at Dependabot? This is wonderful to keep projects up to date. Have a look in the project's "Settings > Code security and analysis"

**Comment by maxandersen** [comment]
@jdubois why not at least test graal on java 17?

**Comment by langchain4j** [comment]
@jdubois you can [fix](https://github.com/mockito/mockito/issues/3143) this problem by adding an updated version of byte-buddy to langchain4j-parent pom.xml <dependencyManagement> section:
```
            <!-- An updated version of byte-buddy is required for mockito to work on Java 21-->
            <dependency>
                <groupId>net.bytebuddy</groupId>
                <artifactId>byte-buddy</artifactId>
                <version>1.14.10</version>
            </dependency>
```
This way we can keep java 8 compatibility.

**Comment by langchain4j** [comment]
@jdubois thanks for the tip with dependabot, I have enabled it

**Comment by jdubois** [comment]
Oh thanks @langchain4j I've just pushed a change, to have Java 8 support.

@maxandersen on Java 17, GaalVM currently fails with: 

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.176 s <<< FAILURE! -- in dev.langchain4j.code.graal.GraalPythonExecutionEngineTest
[ERROR] dev.langchain4j.code.graal.GraalPythonExecutionEngineTest.should_execute_code -- Time elapsed: 0.175 s <<< ERROR!
org.graalvm.polyglot.PolyglotException: java.lang.IllegalAccessError: class org.bouncycastle.jce.provider.BouncyCastleProvider (in module org.bouncycastle.provider) cannot access class java.util.logging.Logger (in module java.logging) because module org.bouncycastle.provider does not read module java.logging
	at org.bouncycastle.provider/org.bouncycastle.jce.provider.BouncyCastleProvider.<clinit>(BouncyCastleProvider.java:75)
	at org.graalvm.py/com.oracle.graal.python.builtins.objects.ssl.CertUtils.<clinit>(CertUtils.java:118)
	at org.graalvm.py/com.oracle.graal.python.builtins.modules.hashlib.HashlibModuleBuiltins.<clinit>(HashlibModuleBuiltins.java:141)

**Comment by aminnasiri** [comment]
> Surprisingly a lot of people still use java 11, even java 8. We did an effort to stay compatible with java 8 so that more people can take advantage of this library.

As contributors to open-source projects, we should push people to pass Java-8 behind. I think Java-11 or might be Java-17 should be the minimum base for new open-source projects like Langchain4J.

**Review by ghost — APPROVED** [review]
@jdubois thank you so much again!

**Comment by jdubois** [comment]
Wonderful, thank you!

**Pull request merged**